### ISSUE TITLE
Rewrite the `Packable` trait to make it `no_std` compatible.

### ISSUE THREAD

**Issue by pvdrz** [iotaledger/bee#533]
## Description

Rewrite the `Packable` trait to make it `no_std` compatible.

## Motivation

Right now the `Packable` trait inside `bee-common` depends explicitly on several types from `std::io` whose migration to `core` seems unlikely. 

## Requirements

Write a list of what you want this feature to do.

1. Be able to use `Packable` in a `no_std` crate.
2. Still be able to work with `std::io` if available.
3. Be able to `#[derive(Packable)]` if no additional checks are needed to pack a type. 

## Open questions (optional)

- It seems there is no standard alternative to `std::io` (no pun intended). There is [genio](https://crates.io/crates/genio) but its development stopped and it is being redesigned. Do we want to have our own general purpose `io` library as part of `bee-common`?

- The naive way to get out would be to introduce two new traits `Read` and `Write` and add either an associated type or generic type parameter to `Packable`, `Read` and `Write` for the error:
  -  Using an associated type restricts us to having to provide a single error type for every implementation beforehand. Which would mean we would have to also provide our own readers and writers. All of this conflicts heavily with (2).
  -  On the other hand using a generic type does not have any of those problems but sacrifices usability because you cannot do `true.packed_len()` but have to write `Packable::<E>::packed_len(&true)` instead. It is also more fragile because an user could provide different `packed_len`s for different error types. We could move `packed_len` to another trait `PackedSized` with no generics and make `Packable<E>: PackedSized` as a workaround.

- Another option would be to move away from the fallible IO approach and make all the `Packable` functions infallible. It would be the "packer's" responsibility to provide a `Vec` or slice with enough space and with the appropriate lifetime to pack and unpack the desired value.  
- Is it feasible to provide our own `serde` serializer and deserializer given that it is already `no_std` and remove `Packable` completely?

## Are you planning to do it yourself in a pull request?

Yes

**Comment by pvdrz** [comment]
I've been playing with some ideas involving serde and I think removing the responsibility of handling the buffers used to pack and unpack is a good thing for `no_std` support. 

Even though serde is an awesome crate that can be used to avoid a lot of boilerplate, there is this issue around serializing arrays that haven't been fixed because serde hit 1.0 before const generics were stable. Because of that, serializing large arrays requires extra dependencies like the `serde-big-array` crate. Boxed slices also have some ergonomic issues. It is also true that this entangles `Packable` with any other serialization provided by serde. Which means that we would have to be "consistent"
 in our serialization in case we wanted to implement `Serialize` and `Deserialize` manually for a type.

 That being said, I have a [small PoC](https://github.com/pvdrz/bee/commit/5f32f350c0106c171b7ce03b261384b56fe873e0) to use serde to replace packable ( you can check the workaround for arrays in the tests for packable).

I've been thinking that we could modify the `Packable` trait in such a way that it never involves `std::io` readers and writers but our own `Packer` and `Unpacker` types (akin to `Serializer` and `Deserializer`). And, instead of using the serde [data model](https://serde.rs/data-model.html) we would use our own data model (we don't care about struct's names, arrays are handled properly, size is prefixed for variable-length entities, we don't handle chars, etc).

The `Packer` would provide methods like `pack_u8`, `pack_u16`, ... `pack_array` and so on. And `Packable` would be implemented in a similar fashion to what serde does. Each field of a struct is serialized calling one of those methods. The pros here is that we don't have to assume a specific way of writing the bytes to someplace, the `Packer` implementation would take care of serializing basic types. Additionally we could provide a `#[derive(Packable)]` macro and other custom attribute macros to tweak how enums are packed for example.

**Comment by pvdrz** [comment]
To be concise this is more or less what I have in mind:

First, we delegate the capacity of packing and unpacking bytes to the `Packer` and `Unpacker` traits:

```rust
pub trait Packer {
    type Error: PackError;

    fn pack_bytes(&mut self, bytes: &[u8]) -> Result<(), Self::Error>;
}

pub trait Unpacker {
    type Error: UnpackError;

    fn unpack_exact_bytes<const N: usize>(&mut self) -> Result<&[u8; N], Self::Error>;

    fn unpack_bytes(&mut self, n: usize) -> Result<&[u8], Self::Error>;
}
```

These could be easily implemented for `Vec<u8>` and `&[u8]` and free us from depending on `std::io`. Additionally it would be possible to implement a `Packer` backed by some stack-based memory allocation or by `Allocator` directly if desired.

The associated error types are bounded by two traits in case we need some specific errors to be raised. 

Now, the `Packable` trait:

```rust
pub trait Packable: Sized {
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error>;
    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error>;
}
```

Nothing special there. The big first win is that implementing `Packable` doesn't restrict us to an specific way of writing or reading bytes as we are generic over `Packer` and `Unpacker` accordingly. Some implementations of this trait are:

```rust
impl Packable for $integer_type {
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        packer.pack_bytes(&self.to_le_bytes())
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
        let bytes: &[u8; size_of::<Self>()] = unpacker.unpack_exact_bytes()?;
        Ok(Self::from_le_bytes(*bytes))
    }
}

impl<T: Packable> Packable for Option<T> {
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        match self {
            Some(item) => {
                1u8.pack(packer)?;
                item.pack(packer)
            }
            None => 0u8.pack(packer),
        }
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
        match u8::unpack(unpacker)? {
            0 => Ok(None),
            1 => Ok(Some(T::unpack(unpacker)?)),
            n => Err(U::Error::invalid_variant(n as u64)),
        }
    }
}

impl<T: Packable> Packable for Vec<T> {
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        (self.len() as u64).pack(packer)?;

        for item in self.iter() {
            item.pack(packer)?;
        }

        Ok(())
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
        let len = u64::unpack(unpacker)? as usize;

        let mut vec = Vec::with_capacity(len);

        for _ in 0..len {
            let item = T::unpack(unpacker)?;
            vec.push(item);
        }

        Ok(vec)
    }
}
```

Packing arrays without heap allocations is possible thanks to `std::mem::MaybeUninit`:

```rust
impl<T: Packable, const N: usize> Packable for [T; N] {
    fn pack<P: Packer>(&self, packer: &mut P) -> Result<(), P::Error> {
        for item in self.iter() {
            item.pack(packer)?;
        }

        Ok(())
    }

    fn unpack<U: Unpacker>(unpacker: &mut U) -> Result<Self, U::Error> {
        let mut array: [MaybeUninit<T>; N] =
            unsafe { MaybeUninit::<[MaybeUninit<T>; N]>::uninit().assume_init() };

        for item in array.iter_mut() {
            let unpacked = T::unpack(unpacker)?;
            unsafe {
                item.as_mut_ptr().write(unpacked);
            }
        }

        Ok(unsafe { (&array as *const _ as *const Self).read() })
    }
}
```

and finally some tests :)

```rust
#[cfg(test)]
mod tests {
    use super::*;

    enum Error {
        UnexpectedEOF,
        InvalidVariant(u64),
    }

    impl PackError for Error {}

    impl UnpackError for Error {
        fn invalid_variant(value: u64) -> Self {
            Self::InvalidVariant(value)
        }
    }

    #[derive(Default)]
    struct VecPacker {
        vec: Vec<u8>,
    }

    impl Packer for VecPacker {
        type Error = Error;

        fn pack_bytes(&mut self, bytes: &[u8]) -> Result<(), Self::Error> {
            self.vec.extend_from_slice(bytes);
            Ok(())
        }
    }

    struct SliceUnpacker<'un> {
        slice: &'un [u8],
    }

    impl<'un> SliceUnpacker<'un> {
        fn new(slice: &'un [u8]) -> Self {
            Self { slice }
        }
    }

    impl<'un> Unpacker for SliceUnpacker<'un> {
        type Error = Error;

        fn unpack_exact_bytes<const N: usize>(&mut self) -> Result<&[u8; N], Self::Error> {
            Ok(unsafe { &*(self.unpack_bytes(N)? as *const [u8] as *const [u8; N]) })
        }

        fn unpack_bytes(&mut self, n: usize) -> Result<&[u8], Self::Error> {
            let (head, tail) = self.slice.split_at(n);

            if head.len() == n {
                self.slice = tail;
                Ok(head)
            } else {
                Err(Error::UnexpectedEOF)
            }
        }
    }

    #[test]
    fn vec_of_vecs() {
        let value = vec![vec![1u64], vec![1, 2], vec![1, 2, 3]];

        let mut packer = VecPacker::default();
        value.pack(&mut packer).ok().unwrap();

        let mut unpacker = SliceUnpacker::new(packer.vec.as_slice());
        let result = Vec::<Vec<u64>>::unpack(&mut unpacker).ok().unwrap();

        assert_eq!(value, result);
    }
}
```

Let me know what you think :)

One final goodie is that we could add a custom derive to make everything easier even for enums

```rust
#[derive(Packable)]
enum Maybe<T> {
    #[packable::variant(69)]
    Just(T),
    #[packable::variant(42)]
    Nothing,
}
```

**Comment by zesterer** [comment]
I'm broadly in favour of this, with the following exception:

> The Packer would provide methods like pack_u8, pack_u16, ... pack_array and so on.

`serde` has methods like this specifically because it supports multiple backends and so can represent this data in multiple ways. Because `Packable` only supports writing bytes, there's no call for the `Packer` trait to be the one to provide this functionality: it can just be something controlled by the implementation `impl Packable for uX`.

I assume that because these methods weren't included in the example implementation, you've also decided against them?

**Comment by pvdrz** [comment]
Yes I realized while writing it that we could just use an implementation of Packable for each integer type

**Comment by pvdrz** [comment]
Solved via #569.
